plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -76,13 +77,77 @@ protected function initialize() {
 		// Extend allowed `collection_params` for the REST API.
 		add_filter( 'rest_product_collection_params', array( $this, 'extend_rest_query_allowed_params' ), 10, 1 );
 
+		// Provide location context into block's context.
+		add_filter( 'render_block_context', array( $this, 'provide_location_context' ), 10, 3 );
+		add_filter( 'render_block_woocommerce/product-template', array( $this, 'test' ), 0, 3 );
+
 		// Interactivity API: Add navigation directives to the product collection block.
 		add_filter( 'render_block_woocommerce/product-collection', array( $this, 'enhance_product_collection_with_interactivity' ), 10, 2 );
 		add_filter( 'render_block_core/query-pagination', array( $this, 'add_navigation_link_directives' ), 10, 3 );
 
 		add_filter( 'posts_clauses', array( $this, 'add_price_range_filter_posts_clauses' ), 10, 2 );
 	}
 
+	// TODO: Remove this method.
+	public function test( $block_content, $block, $instance ) {
+		$location = $instance->context['location'] ?? array();
+		$block_content = print_r( $location, true ) . $block_content;
+		return $block_content;
+	}
+
+	/**
+	 * Provides the location context into the block's context.
+	 *
+	 * The location context's schema is an array with the following:
+	 * - type: The context type. Possible values are 'site', 'order', 'cart', 'archive', 'product'.
+	 * - sourceData: The context data.
+	 *
+	 * The sourceData structure depends on the context type as follows:
+	 * - site: []
+	 * - order: ['orderId' => int]
+	 * - cart: ['productIds' => array]
+	 * - archive: ['taxonomy' => string, 'termId' => int]
+	 * - product: ['productId' => int]
+	 *
+	 * @param array    $context      The block context.
+	 * @param array    $parsed_block The parsed block.
+	 * @param WP_Block $parent_block The parent block.
+	 *
+	 * @return array The block's context including the location context.
+	 */
+	public function provide_location_context( $context, $parsed_block, $parent_block ) {
+
+		// Run only on frontend.
+		// This is needed to avoid SSR renders within patterns. @see https://github.com/woocommerce/woocommerce/issues/45181
+		if ( is_admin() || \WC()->is_rest_api_request() ) {
+			return $context;
+		}
+
+		// Only on Product Collection's children blocks.
+		if ( ! is_a( $parent_block, 'WP_Block' ) || 'woocommerce/product-collection' !== $parent_block->name ) {
+			return $context;
+		}
+
+		// Parse ancestor-level context.
+		// @see Automattic\WooCommerce\Blocks\BlockTypes\SingleProduct::update_context()
+		$is_in_single_product = ! empty( $parent_block->context['singleProduct'] ) && $parent_block->context['singleProduct'] && ! empty( $parent_block->context['postId'] );
```

Comment:
> `$parent_block->context['singleProduct']`

Interesting, how this context is added by Single Product block [here](https://github.com/woocommerce/woocommerce/blob/399f9576c2ad963c287c79096413a6025ea82e9d/plugins/woocommerce/src/Blocks/BlockTypes/SingleProduct.php-L149). I am just wondering if it would make sense to use `$parent_block->available_context` instead of `$parent_block->context`?

- Because `$parent_block->available_context` contains all the context provided by ancestor blocks. What do you think?

Reply:
Shared some thoughts https://github.com/woocommerce/woocommerce/pull/44145#discussion_r1513721101